File: it-agents/skills/code-architect/references/code-examples/python/builder.py
"""Builder — construct complex objects step by step."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field


@dataclass
class Document:
    sections: list[str] = field(default_factory=list)

    def render(self) -> str:
        return "\n".join(self.sections)


class DocumentBuilder(ABC):
    @abstractmethod
    def add_title(self, text: str) -> "DocumentBuilder":
        ...

    @abstractmethod
    def add_paragraph(self, text: str) -> "DocumentBuilder":
        ...

    @abstractmethod
    def build(self) -> Document:
        ...


class PlainTextBuilder(DocumentBuilder):
    def __init__(self) -> None:
        self._doc = Document()

    def add_title(self, text: str) -> "PlainTextBuilder":
        self._doc.sections.append(text.upper())
        return self

    def add_paragraph(self, text: str) -> "PlainTextBuilder":
        self._doc.sections.append(text)
        return self

    def build(self) -> Document:
        doc, self._doc = self._doc, Document()
        return doc


class Director:
    """Encapsulates a reusable construction recipe."""

    def __init__(self, builder: DocumentBuilder) -> None:
        self._builder = builder

    def make_report(self) -> Document:
        return (
            self._builder
            .add_title("Quarterly Report")
            .add_paragraph("Revenue grew steadily.")
            .build()
        )


if __name__ == "__main__":
    director = Director(PlainTextBuilder())
    print(director.make_report().render())

File: it-agents/skills/code-architect/references/code-examples/swift/Builder.swift
// Builder — construct complex objects step by step.

struct Report {
    var title: String = ""
    var sections: [String] = []
    var footer: String = ""
}

protocol ReportBuilder {
    func setTitle(_ text: String) -> Self
    func addSection(_ text: String) -> Self
    func setFooter(_ text: String) -> Self
    func build() -> Report
}

final class TextReportBuilder: ReportBuilder {
    private var report = Report()

    @discardableResult
    func setTitle(_ text: String) -> Self {
        report.title = text
        return self
    }

    @discardableResult
    func addSection(_ text: String) -> Self {
        report.sections.append(text)
        return self
    }

    @discardableResult
    func setFooter(_ text: String) -> Self {
        report.footer = text
        return self
    }

    func build() -> Report { report }
}

// Director encapsulates a reusable construction recipe.
struct ReportDirector {
    func assembleSummary(with builder: ReportBuilder) -> Report {
        builder
            .setTitle("Quarterly Summary")
            .addSection("Revenue grew across all regions.")
            .addSection("Costs remained stable.")
            .setFooter("— Confidential —")
            .build()
    }
}

func runDemo() {
    let report = ReportDirector().assembleSummary(with: TextReportBuilder())
    print(report.title)
    report.sections.forEach { print("  • \($0)") }
    print(report.footer)
}

runDemo()
